optimize_workspace.py

```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env python3
+
+# Build script for cargo workspaces
+
+CARGO_PATH="cargo"
+PACKAGE_PREFIX="contracts/"
+
+import glob
+import os
+import stat
+import subprocess
+import toml
+
+def log(*args):
+    print(*args, flush=True)
+
+with open("Cargo.toml") as file:
+    document = toml.load(file)
+    members = document['workspace']['members']
+
+log("Found workspace member entries:", members)
+
+all_packages = []
+for member in members:
+    all_packages.extend(glob.glob(member))
+all_packages.sort()
+log("Package directories:", all_packages)
+
+contract_packages = [p for p in all_packages if p.startswith(PACKAGE_PREFIX)]
```

Comment:
I'd see this as a transition from searching `Cargo.toml`s to parsing and understanding the workspace. The current path filter is probably not ideal. Maybe we end up with a way to filter contracts from non-contracts using a more stable way.

But this is already an improvement over path search since you can specify `members = ["contracts/a", "contracts/b", "packages/*"]` and then `contracts/c` will not be built since it is not in the workspace.